device/device.go

```diff
@@ -86,80 +86,87 @@ func newDevice(deviceID, realm, credentialsSecret string, pairingBaseURL string,
 // Connect connects the device through a goroutine
 func (d *Device) Connect(result chan<- error) {
 	go func(result chan<- error) {
+		// Let's check the channel is valid
+		if result == nil {
+			fmt.Println("The channel cannot be nil.")
+			return
+		}
+
 		// Are we connected already?
 		if d.IsConnected() {
-			if result != nil {
-				result <- nil
-			}
+			result <- nil
 			return
 		}
 
 		// At least one interface available?
 		if len(d.interfaces) == 0 {
-			if result != nil {
-				result <- errors.New("Add at least an interface before attempting to connect")
-			}
+			result <- errors.New("Add at least an interface before attempting to connect")
 			return
 		}
 
-		// First of all, ensure we have a certificate
-		if err := d.ensureCertificate(); err != nil {
-			if result != nil {
-				result <- err
+		// First of all, get the broker URL with an HTTP request
+		brokerURL, err := d.getBrokerURL()
+		for {
+			if err != nil && d.AutoReconnect {
```

Comment:
This should be handled through an exponential backoff (https://github.com/cenkalti/backoff) and most of all, we should give the option to configure the overall timeout for the Connect process